### PR TITLE
feat(ENG-6845): return `cart`, `userErrors` and `errors` from all client methods

### PR DESCRIPTION
**Pull request by alexcasche**
## Why are these changes introduced?

Addresses [ENG-6845](https://nacelle.atlassian.net/browse/ENG-6845)

> Return `cart`, `userErrors` and `errors` from all client actions

## What is this pull request doing?

- Creates a new return type called `CartResponse`
- Creates a utility function called `formatCartResponse` to remove duplicate code across actions, and simplify testing
- Creates new tests, and updates existing tests to work with new format

## How to Test

1. Check out the `ENG-6845/return-cart-errors-userErrors` branch.
2. `npm run bootstrap` from the monorepo root, as needed.
3. `cd packages/shopify-cart`
4. `npm run test:ci` - tests should pass with expected coverage

Screen Shot 2022-08-29 at 12 48 09 PM

### CODE REVIEW

**Review by KrisQ — APPROVED** [review]
Looks good & works great!

Screen Shot 2022-08-29 at 3 03 16 PM

File: packages/shopify-cart/src/client/actions/cartAttributesUpdate.ts [hunk 1]
```diff
@@ -37,17 +37,11 @@ export default async function cartNoteUpdate({
...
-
-    const cart = cartResponse.data?.cartAttributesUpdate.cart;
-
-    if (cart) {
-      return cartFromGql({ cart });
-    }
+    return formatCartResponse({
+      cart: shopifyResponse.data?.cartAttributesUpdate.cart,
+      userErrors: shopifyResponse.data?.cartAttributesUpdate.userErrors,
+      errors: shopifyResponse?.errors
```
Comment (NWRichmond) [inline packages/shopify-cart/src/client/actions/cartAttributesUpdate.ts:43]:
```suggestion
      errors: shopifyResponse.errors
```

File: packages/shopify-cart/src/client/actions/cart.ts [hunk 2]
```diff
@@ -10,33 +10,31 @@ export interface CartParams {
...
-
-    const cartData = cartResponse.data?.cart;
-
-    if (cartData) {
-      return cartFromGql({ cart: cartData });
-    }
+    return formatCartResponse({
+      cart: shopifyResponse.data?.cart,
+      userErrors: null,
+      errors: shopifyResponse?.errors
```
Comment (NWRichmond) [inline packages/shopify-cart/src/client/actions/cart.ts:36]:
```suggestion
      errors: shopifyResponse.errors
```

File: packages/shopify-cart/src/client/actions/cartBuyerIdentityUpdate.ts [hunk 3]
```diff
@@ -37,17 +37,11 @@ export default async function CartBuyerIdentityUpdate({
...
-
-    const cart = cartResponse.data?.cartBuyerIdentityUpdate.cart;
-
-    if (cart) {
-      return cartFromGql({ cart });
-    }
+    return formatCartResponse({
+      cart: shopifyResponse.data?.cartBuyerIdentityUpdate.cart,
+      userErrors: shopifyResponse.data?.cartBuyerIdentityUpdate.userErrors,
+      errors: shopifyResponse?.errors
```
Comment (NWRichmond) [inline packages/shopify-cart/src/client/actions/cartBuyerIdentityUpdate.ts:43]:
```suggestion
      errors: shopifyResponse.errors
```

File: packages/shopify-cart/src/client/actions/cartCreate.ts [hunk 4]
```diff
@@ -34,17 +34,11 @@ export default async function cartCreate({
...
-
-    const cart = cartResponse.data?.cartCreate.cart;
-
-    if (cart) {
-      return cartFromGql({ cart });
-    }
+    return formatCartResponse({
+      cart: shopifyResponse.data?.cartCreate.cart,
+      userErrors: shopifyResponse.data?.cartCreate.userErrors,
+      errors: shopifyResponse?.errors
```
Comment (NWRichmond) [inline packages/shopify-cart/src/client/actions/cartCreate.ts:40]:
```suggestion
      errors: shopifyResponse.errors
```

File: packages/shopify-cart/src/client/actions/cartDiscountCodesUpdate.ts [hunk 5]
```diff
@@ -36,17 +36,11 @@ export default async function cartDiscountCodesUpdate({
...
-
-    const cart = cartResponse.data?.cartDiscountCodesUpdate.cart;
-
-    if (cart) {
-      return cartFromGql({ cart });
-    }
+    return formatCartResponse({
+      cart: shopifyResponse.data?.cartDiscountCodesUpdate.cart,
+      userErrors: shopifyResponse.data?.cartDiscountCodesUpdate.userErrors,
+      errors: shopifyResponse?.errors
```
Comment (NWRichmond) [inline packages/shopify-cart/src/client/actions/cartDiscountCodesUpdate.ts:42]:
```suggestion
      errors: shopifyResponse.errors
```

File: packages/shopify-cart/src/client/actions/cartLinesAdd.ts [hunk 6]
```diff
@@ -36,17 +36,11 @@ export default async function cartLinesAdd({
...
-
-    const cart = cartResponse.data?.cartLinesAdd.cart;
-
-    if (cart) {
-      return cartFromGql({ cart });
-    }
+    return formatCartResponse({
+      cart: shopifyResponse.data?.cartLinesAdd.cart,
+      userErrors: shopifyResponse.data?.cartLinesAdd.userErrors,
+      errors: shopifyResponse?.errors
```
Comment (NWRichmond) [inline packages/shopify-cart/src/client/actions/cartLinesAdd.ts:42]:
```suggestion
      errors: shopifyResponse.errors
```

File: packages/shopify-cart/src/client/actions/cartLinesRemove.ts [hunk 7]
```diff
@@ -36,17 +36,11 @@ export default async function cartLinesRemove({
...
-
-    const cart = cartResponse.data?.cartLinesRemove.cart;
-
-    if (cart) {
-      return cartFromGql({ cart });
-    }
+    return formatCartResponse({
+      cart: shopifyResponse.data?.cartLinesRemove.cart,
+      userErrors: shopifyResponse.data?.cartLinesRemove.userErrors,
+      errors: shopifyResponse?.errors
```
Comment (NWRichmond) [inline packages/shopify-cart/src/client/actions/cartLinesRemove.ts:42]:
```suggestion
      errors: shopifyResponse.errors
```

File: packages/shopify-cart/src/client/actions/cartLinesUpdate.ts [hunk 8]
```diff
@@ -37,17 +37,11 @@ export default async function cartLinesUpdate({
...
-
-    const cart = cartResponse.data?.cartLinesUpdate.cart;
-
-    if (cart) {
-      return cartFromGql({ cart });
-    }
+    return formatCartResponse({
+      cart: shopifyResponse.data?.cartLinesUpdate.cart,
+      userErrors: shopifyResponse.data?.cartLinesUpdate.userErrors,
+      errors: shopifyResponse?.errors
```
Comment (NWRichmond) [inline packages/shopify-cart/src/client/actions/cartLinesUpdate.ts:43]:
```suggestion
      errors: shopifyResponse.errors
```

File: packages/shopify-cart/src/client/actions/cartNoteUpdate.ts [hunk 9]
```diff
@@ -36,17 +36,11 @@ export default async function cartNoteUpdate({
...
-
-    const cart = cartResponse.data?.cartNoteUpdate.cart;
-
-    if (cart) {
-      return cartFromGql({ cart });
-    }
+    return formatCartResponse({
+      cart: shopifyResponse.data?.cartNoteUpdate.cart,
+      userErrors: shopifyResponse.data?.cartNoteUpdate.userErrors,
+      errors: shopifyResponse?.errors
```
Comment (NWRichmond) [inline packages/shopify-cart/src/client/actions/cartNoteUpdate.ts:42]:
```suggestion
      errors: shopifyResponse.errors
```

File: packages/shopify-cart/src/utils/formatCartResponse.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  userErrors,
+  errors
+}: FormatCartResponseParams): CartResponse {
+  const cartResponse: CartResponse = {
+    cart: null,
+    userErrors: null,
+    errors: null
+  };
+
+  cartResponse.errors = errors ?? null;
```
Comment (NWRichmond) [inline packages/shopify-cart/src/utils/formatCartResponse.ts:28]: Not a big deal at all, but might be worth considering using the same pattern for `errors` as is being used for `userErrors` & `cart`.

File: packages/shopify-cart/src/utils/cartFromGql.ts [hunk 11]
```diff
@@ -6,7 +6,7 @@ export interface CartFromGqlParams {
 }
 
 /**
- * Conver a `cart` from a Storefront GraphQL response to a `Cart`
+ * Convert a `cart` from a Storefront GraphQL response to a `Cart`
```
Comment (NWRichmond) [inline packages/shopify-cart/src/utils/cartFromGql.ts:9]: Good catch!

**Review by NWRichmond — APPROVED** [review]
Nice! All of the response handling matches the names of the query & mutations, and all tests are passing.

The type name `CartResponse` seems very sensible.

I made some suggestions, but nothing crucial that would require a re-review.

File: packages/shopify-cart/src/utils/formatCartResponse.ts (hunk 10 above)
Comment (alexcasche) [reply packages/shopify-cart/src/utils/formatCartResponse.ts]: All set

**Pull request merged**